Route Abstraction

theosiemensrhodes opened the pull request:
Updated how routes are managed and assigned. Most routes/controllers were changed and to remove now unneeded validation.

## How the new Routes are managed

Routes now consist of object schema like this small sample
```typescript
export const AdminRoutes: RouteDefinition = {
    path: '/admin',
    middleware: [
        isAuthorized,
        isAdmin,
    ],
    children: [
        {
            path: '/unverified-volunteers',
            method: 'post',
            action: getUnverifiedVolunteers
        },
        {
            path: '/verify-volunteer',
            method: 'post',
            validation: [
                body('volunteer_id').isUUID('4')
            ],
            action: verifyVolunteer
        },
    ]
};
```

Routes are made up of nested `RouteDefinition` objects where are either `RouteGroup` and `RouteEndpoint` schema as defined in `utils/types.ts`.
```typescript
export type HTTPMethod = 'get' | 'post' | 'put' | 'delete' | 'patch' | 'options' | 'head';

export interface RouteGroup {
    path: string;
    validation?: ValidationChain[];
    middleware?: Array<(req: Request, res: Response, next: NextFunction) => void>;
    children: RouteDefinition[];
}

export interface RouteEndpoint {
    path: string;
    validation?: ValidationChain[];
    middleware?: Array<(req: Request, res: Response, next: NextFunction) => void>;
    method: HTTPMethod;
    action: (req: Request, res: Response, next: NextFunction) => Promise<any>;
}

export type RouteDefinition = RouteGroup | RouteEndpoint;
```
Groups are used if many endpoints share a path, like all the endpoints that sit under `/classes/*`. Endpoints define an endpoint itself, specifically the HTTP method and action associated with the endpoint.

Both Groups and Endpoint can have middleware and validation. The order of which they are applied is from top level to bottom and in each array from first to last i.e.
```typescript
export const Routes: RouteDefinition = {
    path: '/path',
    middleware: [
        first,
        second,
    ],
    children: [
        {
            path: '/path2',
            method: 'post',
            middleware: [
                third,
                etc...
            ],
            action: getUnverifiedVolunteers
        },
    ]
};
```
Middleware and validation applied to groups will then be applied to all children of this group. The route registering function is also in `routes/types.ts` for anyone curious.

## Validation

The validation is handles by the express-validator library. Validation is pretty easy to understand by looking but heres a link to the [express-validator docs](https://express-validator.github.io/docs/) as well for any questions.

@jansm04 it would be great if you could double check the routes you just added. I had a little bit of a merge conflict and I want to make sure I didnt break anything you added.

jjessieshang reviewed (approved):
This is sooo clean!! I love the abstraction of bulky validators in the controllers to the routes now. Also all the validator choices of optional or not, min values, type checks, that you added make it so much easier for devs to understand endpoints written by others. Looks great Theo!

jansm04 reviewed (commented):
Wow that is a massive improvement, great work man. All endpoints I added look good. Nice catch updating the code in VolunteerDetailsCard.ts as well

Just curious about one thing, how come some functions in controllers don't use the express-validator library? (Ex. in the shift module you have the updated validation for addShift() but not for getShiftInfo(), even though they require similar fields in the req body)

JoshFung reviewed (approved):
Wow Theo this is an insanely cool refactor! I feel like I'm always learning about a new library through your PRs and I can't wait for this to be merged into `main`.

Looks good to me for now, but will definitely need to try out working on some routes myself to fully understand it all ! 😄

theosiemensrhodes commented:
> Wow that is a massive improvement, great work man. All endpoints I added look good. Nice catch updating the code in VolunteerDetailsCard.ts as well
> 
> Just curious about one thing, how come some functions in controllers don't use the express-validator library? (Ex. in the shift module you have the updated validation for addShift() but not for getShiftInfo(), even though they require similar fields in the req body)

Good catch Jan, not sure if it was a merge conflict but there were a few spots in the controllers which still had left over validation. Fixed now, I did one more time over to ensure everything was merged well

jansm04 reviewed (approved):
Sweet, looks good to go!
